### PR TITLE
Tryout spatial model

### PR DESCRIPTION
**Pull request by twallema**
* [x] I have checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change
* [x] I made sure all new and existing data sets have a description in the `data/README.md` file
* [x] I have updated the documentation accordingly. (--> models.md on documentation website was updated)

Describe your fixes/additions/changes

## Spatial modeling framework

Use `notebooks/scratch/0.1-twallema-spatial-stochastic.ipynb` as demo.

### What?

The spatial framework builds upon the stochastic national-level model to take spatial heterogeneity into account. The Belgian territory is divided into 43 arrondissements, which are from hereon referred to as *patches*. Our patch-model boils down to a simulation of the extended SEIRD model dynamics on each patch, where the patches are connected by commuting traffic (individuals going to work). This 'connection' is a 43x43 from-to commuting matrix, named `NIS`, extracted from the 2011 Belgian census data (see data readme). The system of equations is identical to the national level stochastic model `COVID19_SEIRD_sto`, differing only in the chance of infection.

If the individual works within his own arrondissement, the individual is assumed to have all his human-to-human contacts with people from within his home arrondissement only. The interaction matrix for individuals working in their residence arrondissement is the sum of home, school, work, leisure, and other human-to-human interactions per day. If the individual does not work in his home patch, the work and home, school, leisure, other human-to-human contacts must be separated.

I updated the model section on our documentation website with an extra paragraph `Stochastic spatial framework`. This section contains the necessary math behind the spatial model, in the hope to make this more tangible. I would happily include all necessary maths here but GitHub does not seem to support LaTeX in its markdown compilations :'( . To build the docs,

- Checkout to my branch `tryout_recurrent_mobility`
- Compile the docs using `python setup.py build_sphinx` (ubuntu 18)
- Open the html files in the `build` dir

#### Generalizing the stratification concept

So far, only age-stratification was used. The concept was generalized to n-stratifications by several alterations to the functions handling input checks and the function assigning model output to xarray (see changes in `base.py`). The names of the stratified parameters and the names of the stratification must be defined as a list,

```
parameters_stratified_names = [None, ['s','a','h', 'c', 'm_C','m_ICU']]
stratification = ['NIS','Nc']
```

In this example, the spatial axis is named after the Belgian `NIS` codes and has no associated stratified parameters. The age axis is named after the interaction matrix `Nc` and has several associated age-stratified parameters. I verified that both non-spatial models still work.

#### Assigning custom coordinates to a stratification

In the non-spatial, age-stratified models used before, the coordinates of the age axis where simply taken as `np.arange(self.stratification.shape[0])` or `[0 1 2 3 4 5 6 7 8]`. However, for the spatial axis, the best coordinates are the arrondissements NIS codes. When building a model, the user can now provide the optional argument 'coordinates', which assigns the list of NIS codes as coordinates of the spatial (NIS) axis. In the example below, no custom coordinates are assigned to the age axis, implying the default `[0 1 2 3 4 5 6 7 8]` is retained.

```
coordinates=[[11000, 12000, 13000, 21000, 23000, 24000, 25000, 31000, 32000,
       33000, 34000, 35000, 36000, 37000, 38000, 41000, 42000, 43000,
       44000, 45000, 46000, 51000, 52000, 53000, 54000, 55000, 56000,
       57000, 61000, 62000, 63000, 64000, 71000, 72000, 73000, 81000,
       82000, 83000, 84000, 85000, 91000, 92000, 93000],None]
stratification = ['NIS','Nc']
```
#### name2nis (`models/utils.py`)

A simple function to convert a municipality/province/arrondissement name to its respective NIS code. Case insensitive. Do note that looking up an arrondissement requires you to write `Arrondissement Antwerpen`. The list of all available NIS-name combinations can be found in the following pickle `data/interim/census_2011/NIS_name.pkl`.

#### Selecting output

`name2nis` was made to make the selection of model output from the spatial axis (coordinate is NIS code) more easy,
```
out = model.sim(250)
desired_out = out.sel(NIS=name2nis('arrondissement antwerpen')).sum(dim='Nc')['M']
```
will simulate the model and return the total number of individuals experiencing mild symptoms (model state `M`), summed over all age groups, in arrondissement antwerpen only.

#### The coordinates for the spatial axis (the NIS codes of the arrondissements) are hardcoded

My first attempt was to extract the coordinates from one of pickled dataframes I made,

```
initN_df=pd.read_pickle('../../data/interim/census_2011/initN.pkl')
coordinates = initN_df.index.values
```

But when changing the notebook path, the directory of `initN.pkl` must be changed too. Is it possible to have python "find" the desired pickle in our repo?

#### Error in visualisation functions `infected()` and `population_status()` when using the non-spatial model

When asking Python to return the number of dimensions of the xarray output after running the non-spatial model,
```
['Nc', 'time']
```
So clearly there are two dimensions, as should be the case for the non-spatial model. Using the visualization functions `infected()` and `population_status()` returns an identical error, stating the xarray output has more than two dimensions,

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-630195eacf29> in <module>
----> 1 population_status(model.sim(200))

~/Documents/PhD/Corona/Python/git/COVID19-Model/src/covid19model/visualization/output.py in population_status(data, filename, ax, **kwargs)
     51     # create plot using xarray interface
     52     data2plot = data[["S", "E", "I_total", "R"]].to_array(dim="states")
---> 53     lines = data2plot.plot.line(x='time', hue="states", ax=ax, **kwargs)
     54     ax.set_xlabel('days')
     55     ax.set_ylabel('number of patients')

~/anaconda3/envs/COVID_MODEL/lib/python3.8/site-packages/xarray/plot/plot.py in line(self, *args, **kwargs)
    457     @functools.wraps(line)
    458     def line(self, *args, **kwargs):
--> 459         return line(self._da, *args, **kwargs)
    460 
    461     @functools.wraps(step)

~/anaconda3/envs/COVID_MODEL/lib/python3.8/site-packages/xarray/plot/plot.py in line(darray, row, col, figsize, aspect, size, ax, hue, x, y, xincrease, yincrease, xscale, yscale, xticks, yticks, xlim, ylim, add_legend, _labels, *args, **kwargs)
    280     ndims = len(darray.dims)
    281     if ndims > 2:
--> 282         raise ValueError(
    283             "Line plots are for 1- or 2-dimensional DataArrays. "
    284             "Passed DataArray has {ndims} "

ValueError: Line plots are for 1- or 2-dimensional DataArrays. Passed DataArray has 3 dimensions
```

### CODE REVIEW

**Review by JennaVergeynst — CHANGES_REQUESTED** [review]
Some remarks:
- can you give the commuting matrix a more meaningful name than "NIS" (NIS is the code used for arrondissements). For instance... commuting_matrix?
- issue with hard-coded NIS-codes: maybe just save a list of NIS-codes somewhere, and read it in via a function, so that the path issue is avoided
- error in the visualisation methods occurs in the non-spatial method? so it was already there before? or was it after your changes?
- please don't save any data as .pkl files => this is not a robust format (e.g. with an older pandas version it cannot be read in: I get error messages). Rather just use csv files

**Comment by JennaVergeynst** [comment]
Another suggestion: change folder data/raw/geological to data/raw/GIS or anything more meaningful about the content

**Comment by JennaVergeynst** [comment]
- I think it will be helpful for Michiel to use datetimes for checkpoints and plotting in the examples
- For saving the figures/maps, better use the results-path (e.g. a new folder results/maps)
=> in that respect, maybe we can have a paths-file that can be read in on top of a notebook, that contains absolute paths to certain folders. So that relative paths cannot be broken by moving a notebook, and so that the absolute path only has to be changed in one place if needed

**Review by JennaVergeynst — CHANGES_REQUESTED** [review]
I meant the results folder that is already present on the same level of notebooks... I think it is better to only put notebooks in the notebooks folder, and make a substructure in results with results or figures for different aspects

**Comment by twallema** [comment]
> I meant the results folder that is already present on the same level of notebooks... I think it is better to only put notebooks in the notebooks folder, and make a substructure in results with results or figures for different aspects

What results folder? Theres binder, data, docs, hpc, notebooks, references, reports, src
My guess is reports is meant for results?

**Comment by JennaVergeynst** [comment]
> What results folder? Theres binder, data, docs, hpc, notebooks, references, reports, src
> My guess is reports is meant for results?

This is also a results folder currently :-) check the current upstream master

**Review by JennaVergeynst — COMMENTED** [review]
Ah sorry, I thought you had named the matrix itself "NIS". I don't have anything against NIS in the code, where it really points to the NIS-code... Just confused by this sentence: "This 'connection' is a 43x43 from-to commuting matrix, named NIS, extracted from the 2011 Belgian census data (see data readme".

**Comment by twallema** [comment]
I opted to rename `NIS` --> `place` instead of `recurrent_mobility` because the output handling is more intuitive, to select the mildly infected in antwerpen use,
```
out.sel(place=name2nis('arrondissement antwerpen'))['M']
```

**Comment by JennaVergeynst** [comment]
@twallema : Before we merge: I just noticed that the file NIS_name.csv contains old nis-codes (there were changes in 2019). I will replace the file by the correct, most recent NIS-codes

**Comment by JennaVergeynst** [comment]
Also: I would put it in another folder, since it is not related to the census of 2011. I would propose to put it also in a GIS folder (all that has to do with geographical data), so with the structure:
raw
|----GIS
          NIScodes etc.
          |----- shapefiles
OK? Or do you want to keep BE as a subfolder of shapefiles?

File: src/covid19model/models/base.py
Comment (jorisvandenbossche) [inline src/covid19model/models/base.py:417]:
```suggestion
            if self.coordinates and self.coordinates[i] is not None:
                coords.update({self.stratification[i]: self.coordinates[i]})
            else:
                coords.update({self.stratification[i]: np.arange(self.stratification_size[i])})
```

**Comment by JennaVergeynst** [comment]
lastly, I would put these kind of files in "raw", since it is not a result of an analysis...

**Review by jorisvandenbossche — COMMENTED** [review]
Few small comments

**Comment by JennaVergeynst** [comment]
@twallema : can you check the box "allow edit by maintainers" (if it is unchecked)? I cannot push my changes to your branch currently...

File: environment.yml
Comment (twallema) [reply environment.yml:24]: Geopandas asked me kindly for `descartes` and `pyarrow` .

File: data/README.md
Comment (twallema) [reply data/README.md]: All files are now saved in .csv format. Data readme was outdated. Resolved.

**Comment by twallema** [comment]
@JennaVergeynst Can you `0.1-twallema-extract-census-data`

> Few small comments

Thanks @jorisvandenbossche, all issues have been resolved. @JennaVergeynst Do you have any more remarks before merging this change?

**Pull request merged**
